frontend/exporter/src/utils.rs

```diff
@@ -1,57 +1,107 @@
-macro_rules! span_fatal {
-    ($s:ident, $span:expr, $($other:tt)*) => {
-        $s.base().tcx.sess.span_fatal($span,format!($($other)*))
-    }
+macro_rules! format_with_context {
+    ($format_str:expr $(,$arg:expr)* $(; {$($x:expr),*})?) => {
+        format!(
+            concat!(
+                $format_str
+                $(, "\n\nContext:\n", $(concat!(" - ", stringify!($x), ": "), "{:#?}", "\n",)*)?
+            ),
+            $($arg,)*
+            $($($x,)*)?
+        )
+    };
+    ($($tt:tt)*) => {format!($($tt)*)};
 }
 
-/*
-Not useful for now.
-
-macro_rules! no_span_fatal {
-    ($s:ident, $($other:tt)*) => {
-        $s.base().tcx.sess.fatal(format!($($other)*))
+mod internal_helpers {
+    macro_rules! _verb {
+        (fatal, $o:expr, $message:expr) => {
+            $o.struct_fatal($message)
+        };
+        (error, $o:expr, $message:expr) => {
+            $o.struct_err($message)
+        };
+        (warn, $o:expr, $message:expr) => {
+            $o.struct_warn($message)
+        };
     }
-}
-pub(crate) use no_span_fatal;
-*/
-
-macro_rules! fatal {
-    ($s:ident, $($other:tt)*) => {
-        {let tcx = $s.base().tcx;
-        match $s.base().opt_def_id {
-            Some(did) =>
-                tcx.sess.span_fatal(tcx.def_span(did), format!($($other)*)),
-            None =>
-                tcx.sess.fatal(format!($($other)*)),
-        }}
+    macro_rules! _span_verb_base {
+        ($verb:ident, $s:ident, $span:expr, $message:expr) => {{
+            let mut builder = $crate::utils::_verb!($verb, $s.base().tcx.sess, $message);
+            if let Some(span) = $span {
+                builder.set_span(span.clone());
+            }
+            builder.code(rustc_errors::DiagnosticId::Error(format!("HaxFront")));
+            builder.note(
+                "⚠️ This is a bug in Hax's frontend.
+Please report this error with some context (e.g. the current crate)!",
+            );
+            builder.emit()
+        }};
     }
+
+    pub(crate) use _span_verb_base;
+    pub(crate) use _verb;
 }
 
-macro_rules! supposely_unreachable {
-    (@$verb:tt $label:literal : $($e:expr),*$(,)?) => {
-        $verb !(concat!("
-Supposely unreachable place in the Rust AST. The label is ", stringify!($label), ".
-This ", stringify!($verb), " happend because I thought some node in the Rust AST could not show up under certain condition(s).
+macro_rules! report {
+    ($verb:ident, $s:ident [$span:expr], $($tt:tt)*) => {
+        $crate::utils::_span_verb_base!($verb, $s, Some($span), $crate::utils::format_with_context!($($tt)*))
+    };
+    ($verb:ident, $s:ident, $($tt:tt)*) => {
+        $crate::utils::_span_verb_base!(
+            $verb,
+            $s,
+            $s.base().opt_def_id.map(|did| $s.base().tcx.def_span(did)),
+            $crate::utils::format_with_context!($($tt)*)
+        )
+    };
+}
 
-Please report that message along with the crate!
+#[allow(unused_macros)]
```

Comment:
Are these all unused or why do we need the `allow`s here?

Reply:
Only `warning` is unused, actually, removing those other superfluous `allow`s!